File: np_exiv2/lib/src/np_exiv2.dart
import 'dart:convert';
import 'dart:ffi';
import 'dart:io';
import 'dart:isolate';
import 'dart:typed_data';

import 'package:euc/jis.dart';
import 'package:ffi/ffi.dart' as ffi;
import 'package:logging/logging.dart';
import 'package:np_common/object_util.dart';
import 'package:quiver/iterables.dart';
import 'package:time_machine2/time_machine2.dart';

import 'np_exiv2_bindings_generated.dart';

enum TypeId {
  unsignedByte,
  asciiString,
  unsignedShort,
  unsignedLong,
  unsignedRational,
  signedByte,
  undefined,
  signedShort,
  signedLong,
  signedRational,
  tiffFloat,
  tiffDouble,
  tiffIfd,
  unsignedLongLong,
  signedLongLong,
  tiffIfd8,
  string,
  date,
  time,
  comment,
  directory,
  xmpText,
  xmpAlt,
  xmpBag,
  xmpSeq,
  langAlt,
  invalidTypeId;

  factory TypeId.fromNative(Exiv2TypeId src) {
    return TypeId.values[src.index];
  }
}

class Date {
  const Date(this.year, this.month, this.day);

  final int year;
  final int month;
  final int day;
}

class Time {
  const Time(this.hour, this.minute, this.second, this.tzHour, this.tzMinute);

  final int hour;
  final int minute;
  final int second;
  final int tzHour;
  final int tzMinute;
}

class Rational {
  const Rational(this.numerator, this.denominator);

  double toDouble() => numerator / denominator;

  @override
  String toString() => "$numerator/$denominator";

  final int numerator;
  final int denominator;
}

class Value {
  const Value({
    required this.typeId,
    required Uint8List data,
    required int count,
  }) : _data = data,
       _count = count;

  T as<T>() => asTyped() as T;

  Object asTyped() {
    try {
      switch (typeId) {
        case TypeId.unsignedByte:
          return _listOrValue(_data);
        case TypeId.asciiString:
          // this is supposed to be ascii but vendors are putting utf-8 strings
          return _removeNullTerminator(utf8.decode(_data));
        case TypeId.unsignedShort:
          return _listOrValue(_data.buffer.asUint16List());
        case TypeId.unsignedLong:
        case TypeId.tiffIfd:
          return _listOrValue(_data.buffer.asUint32List());
        case TypeId.unsignedRational:
          return _listOrValue(
            partition(
              _data.buffer.asUint32List(),
              2,
            ).map((e) => Rational(e[0], e[1])).toList(),
          );
        case TypeId.signedByte:
          return _listOrValue(_data.buffer.asInt8List());
        case TypeId.signedShort:
          return _listOrValue(_data.buffer.asInt16List());
        case TypeId.signedLong:
          return _listOrValue(_data.buffer.asInt32List());
        case TypeId.signedRational:
          return _listOrValue(
            partition(
              _data.buffer.asInt32List(),
              2,
            ).map((e) => Rational(e[0], e[1])).toList(),
          );
        case TypeId.tiffFloat:
          return _listOrValue(_data.buffer.asFloat32List());
        case TypeId.tiffDouble:
          return _listOrValue(_data.buffer.asFloat64List());
        case TypeId.unsignedLongLong:
        case TypeId.tiffIfd8:
          return _listOrValue(_data.buffer.asUint64List());
        case TypeId.signedLongLong:
          return _listOrValue(_data.buffer.asInt64List());
        case TypeId.string:
          return _removeNullTerminator(utf8.decode(_data));
        case TypeId.date:
          return _data.buffer.asInt32List().let((e) => Date(e[0], e[1], e[2]));
        case TypeId.time:
          return _data.buffer.asInt32List().let(
            (e) => Time(e[0], e[1], e[2], e[3], e[4]),
          );
        case TypeId.comment:
          return _convertCommentValue();
        case TypeId.undefined:
        case TypeId.directory:
        case TypeId.invalidTypeId:
          return _data.buffer.asUint8List();
        case TypeId.xmpText:
          return _removeNullTerminator(utf8.decode(_data));
        case TypeId.xmpSeq:
        case TypeId.xmpBag:
          return _removeNullTerminator(utf8.decode(_data)).split(", ");
        case TypeId.xmpAlt:
        case TypeId.langAlt:
          throw UnsupportedError("XMP not fully supported");
      }
    } catch (e, stackTrace) {
      _log.severe(
        "[asTyped] Failed to convert data to type: $typeId, $_data",
        e,
        stackTrace,
      );
      rethrow;
    }
  }

  String toDebugString() {
    return "Value{"
        "typeId: $typeId, "
        "size: ${_data.length}, "
        "count: $_count, "
        "}";
  }

  Object _listOrValue(List<Object> values) {
    return _count == 1 ? values.first : values;
  }

  String _convertCommentValue() {
    // the first 8 chars is the charset, valid values are [ASCII, JIS,
    // UNICODE]
    return _data.buffer.asUint8List().let((e) {
      String? charset;
      Uint8List data = e;
      if (e.length >= 8) {
        charset = ascii.decode(e.sublist(0, 8));
        data = e.sublist(8);
      }
      if (charset == "ASCII") {
        return _removeNullTerminator(ascii.decode(data, allowInvalid: true));
      } else if (charset == "JIS") {
        return _removeNullTerminator(ShiftJIS().decode(data));
      } else if (charset == "UNICODE") {
        // UTF16
        return _removeNullTerminator(
          String.fromCharCodes(data.buffer.asUint16List()),
        );
      } else {
        // unknown, treat as utf8
        return _removeNullTerminator(utf8.decode(data));
      }
    });
  }

  String _removeNullTerminator(String src) {
    // C strings are null terminated, dart's aren't
    if (src.isEmpty) {
      // ?
      return src;
    } else {
      var end = src.length;
      while (end > 0 && src[end - 1] == "\u0000") {
        --end;
      }
      if (end == 0) {
        return "";
      } else {
        return src.substring(0, end);
      }
    }
  }

  final TypeId typeId;
  final Uint8List _data;
  final int _count;

  static final _log = Logger("np_exiv2.Value");
}

class Metadatum {
  const Metadatum({required this.tagKey, required this.value});

  factory Metadatum.fromNative(Exiv2Metadatum src) {
    return Metadatum(
      tagKey: src.tag_key.cast<ffi.Utf8>().toDartString(),
      value: Value(
        typeId: TypeId.fromNative(src.type_id),
        count: src.count,
        data: Uint8List.fromList(src.data.asTypedList(src.size)),
      ),
    );
  }

  final String tagKey;
  final Value value;
}

class ReadResult {
  const ReadResult(
    this.width,
    this.height,
    this.iptcData,
    this.exifData,
    this.xmpData,
  );

  factory ReadResult.fromNative(Exiv2ReadResult src) {
    _log.fine(
      "[fromNative] w: ${src.width}, h: ${src.height}, iptcCount: ${src.iptc_count}, exifCount: ${src.exif_count}, xmpCount: ${src.xmp_count}",
    );
    final iptcData = <Metadatum>[];
    for (var i = 0; i < src.iptc_count; ++i) {
      iptcData.add(Metadatum.fromNative(src.iptc_data[i]));
    }
    final exifData = <Metadatum>[];
    for (var i = 0; i < src.exif_count; ++i) {
      exifData.add(Metadatum.fromNative(src.exif_data[i]));
    }
    final xmpData = <Metadatum>[];
    for (var i = 0; i < src.xmp_count; ++i) {
      xmpData.add(Metadatum.fromNative(src.xmp_data[i]));
    }
    return ReadResult(src.width, src.height, iptcData, exifData, xmpData);
  }

  // always 0 for videos, width of a video can only be retrieved from XMP (if
  // available)
  final int width;
  // always 0 for videos, height of a video can only be retrieved from XMP (if
  // available)
  final int height;
  final List<Metadatum> iptcData;
  final List<Metadatum> exifData;
  final List<Metadatum> xmpData;

  static final _log = Logger("np_exiv2.ReadResult");
}

ReadResult readFile(String path, {required bool isReadXmp}) {
  final stopwatch = Stopwatch()..start();
  final pathC = path.toNativeUtf8();
  try {
    _log.fine("[readFile] Reading $path");
    final result = _bindings.exiv2ReadFile(pathC.cast(), isReadXmp ? 1 : 0);
    if (result == nullptr) {
      _log.severe("[readFile] Result is null for file: $path");
      throw StateError("Failed to read file");
    }
    try {
      return ReadResult.fromNative(result[0]);
    } finally {
      _bindings.exiv2ResultFree(result);
    }
  } finally {
    ffi.malloc.free(pathC);
    _log.fine("[readFile] Done in ${stopwatch.elapsedMilliseconds}ms");
  }
}

ReadResult readBuffer(Uint8List buffer, {required bool isReadXmp}) {
  final stopwatch = Stopwatch()..start();
  Pointer<Uint8>? cbuffer;
  try {
    _log.fine("[readBuffer] Allocating buffer with size: ${buffer.length}");
    cbuffer = ffi.malloc.allocate<Uint8>(buffer.length);
    final cbufferView = cbuffer.asTypedList(buffer.length);
    cbufferView.setAll(0, buffer);
    _log.fine("[readBuffer] Reading buffer");
    final result = _bindings.exiv2ReadBuffer(
      cbuffer,
      buffer.length,
      isReadXmp ? 1 : 0,
    );
    if (result == nullptr) {
      _log.severe("[readBuffer] Result is null for buffer");
      throw StateError("Failed to read buffer");
    }
    try {
      return ReadResult.fromNative(result[0]);
    } finally {
      _bindings.exiv2ResultFree(result);
    }
  } finally {
    if (cbuffer != null) {
      ffi.malloc.free(cbuffer);
      _log.fine("[readBuffer] Done in ${stopwatch.elapsedMilliseconds}ms");
    }
  }
}

Future<ReadResult> readHttp(
  Uri url, {
  Map<String, String>? httpHeaders,
  required bool isReadXmp,
}) {
  final urlStr = url.toString();
  return Isolate.run(() {
    final stopwatch = Stopwatch()..start();
    final headerKeys = <Pointer<Char>>[];
    final headerValues = <Pointer<Char>>[];
    Pointer<Pointer<Char>>? headerKeysC;
    Pointer<Pointer<Char>>? headerValuesC;
    final urlC = urlStr.toNativeUtf8();
    try {
      for (final e in (httpHeaders ?? const {}).entries) {
        headerKeys.add(e.key.toNativeUtf8().cast());
        headerValues.add(e.value.toNativeUtf8().cast());
      }
      headerKeysC = headerKeys.toNativeArray();
      headerValuesC = headerValues.toNativeArray();

      _log.fine("[readHttp] Reading $url");
      final result = _bindings.exiv2ReadHttp(
        urlC.cast(),
        headerKeysC,
        headerValuesC,
        httpHeaders?.length ?? 0,
        isReadXmp ? 1 : 0,
      );
      if (result == nullptr) {
        _log.severe("[readHttp] Result is null for url: $url");
        throw StateError("Failed to read file");
      }
      try {
        return ReadResult.fromNative(result[0]);
      } finally {
        _bindings.exiv2ResultFree(result);
      }
    } finally {
      ffi.malloc.free(urlC);
      for (final e in headerKeys) {
        ffi.malloc.free(e);
      }
      for (final e in headerValues) {
        ffi.malloc.free(e);
      }
      if (headerKeysC != null) {
        ffi.malloc.free(headerKeysC);
      }
      if (headerValuesC != null) {
        ffi.malloc.free(headerValuesC);
      }
      _log.fine("[readHttp] Done in ${stopwatch.elapsedMilliseconds}ms");
    }
  });
}

Future<bool> copyMetadata(
  Uint8List from,
  File to, {
  bool shouldCopyOrientation = true,
  bool shouldCopyThumbnail = true,
}) async {
  try {
    return await _copyMetadata(
      from,
      to.path,
      shouldCopyOrientation: shouldCopyOrientation,
      shouldCopyThumbnail: shouldCopyThumbnail,
    );
  } catch (e, stackTrace) {
    _log.severe("[copyMetadata] Failed while copyMetadata", e, stackTrace);
    return false;
  }
}

Future<bool> writeFileDateTimeOriginal(String path, ZonedDateTime? value) {
  final timeStr = value?.let((e) {
    final dt = e.localDateTime;
    final y = dt.year.toString().padLeft(4, "0");
    final m = dt.monthOfYear.toString().padLeft(2, "0");
    final d = dt.dayOfMonth.toString().padLeft(2, "0");
    final h = dt.hourOfDay.toString().padLeft(2, "0");
    final min = dt.minuteOfHour.toString().padLeft(2, "0");
    final s = dt.secondOfMinute.toString().padLeft(2, "0");
    return "$y:$m:$d $h:$min:$s";
  });
  final offsetStr = value?.let((e) {
    final offsetS = e.offset.inSeconds;
    final sign = offsetS < 0 ? "-" : "+";
    final h = (offsetS.abs() ~/ 3600).toString().padLeft(2, "0");
    final m = ((offsetS.abs() % 3600) ~/ 60).toString().padLeft(2, "0");
    return "$sign$h:$m";
  });
  return Isolate.run(() {
    final pathC = path.toNativeUtf8();
    final timeC = timeStr?.toNativeUtf8();
    final offsetC = offsetStr?.toNativeUtf8();
    try {
      return _bindings.exiv2WriteFileDateTimeOriginal(
            pathC.cast(),
            timeC?.cast() ?? nullptr,
            offsetC?.cast() ?? nullptr,
          ) !=
          0;
    } finally {
      ffi.malloc.free(pathC);
      if (timeC != null) {
        ffi.malloc.free(timeC);
      }
      if (offsetC != null) {
        ffi.malloc.free(offsetC);
      }
    }
  });
}

Future<bool> writeFileGps(
  String path, {
  required String? latitudeRef,
  required List<Rational>? latitude,
  required String? longitudeRef,
  required List<Rational>? longitude,
}) {
  if (!(latitude == null || latitude.length == 3)) {
    throw ArgumentError.value(
      latitude,
      "latitude",
      "Must be null or length == 3",
    );
  }
  if (!(longitude == null || longitude.length == 3)) {
    throw ArgumentError.value(
      longitude,
      "longitude",
      "Must be null or length == 3",
    );
  }
  final latData = latitude
      ?.expand((r) => [r.numerator, r.denominator])
      .toList();
  final lngData = longitude
      ?.expand((r) => [r.numerator, r.denominator])
      .toList();
  return Isolate.run(() {
    final pathC = path.toNativeUtf8();
    final latRefC = latitudeRef?.toNativeUtf8();
    final lngRefC = longitudeRef?.toNativeUtf8();
    Pointer<Uint32>? latC;
    Pointer<Uint32>? lngC;
    try {
      if (latData != null) {
        latC = ffi.malloc.allocate<Uint32>(latData.length * sizeOf<Uint32>());
        for (var i = 0; i < latData.length; ++i) {
          latC[i] = latData[i];
        }
      }
      if (lngData != null) {
        lngC = ffi.malloc.allocate<Uint32>(lngData.length * sizeOf<Uint32>());
        for (var i = 0; i < lngData.length; ++i) {
          lngC[i] = lngData[i];
        }
      }
      return _bindings.exiv2WriteFileGps(
            pathC.cast(),
            latRefC?.cast() ?? nullptr,
            latC ?? nullptr,
            lngRefC?.cast() ?? nullptr,
            lngC ?? nullptr,
          ) !=
          0;
    } finally {
      ffi.malloc.free(pathC);
      if (latRefC != null) {
        ffi.malloc.free(latRefC);
      }
      if (lngRefC != null) {
        ffi.malloc.free(lngRefC);
      }
      if (latC != null) {
        ffi.malloc.free(latC);
      }
      if (lngC != null) {
        ffi.malloc.free(lngC);
      }
    }
  });
}

Future<bool> _copyMetadata(
  Uint8List from,
  String toPath, {
  required bool shouldCopyOrientation,
  required bool shouldCopyThumbnail,
}) {
  return Isolate.run(() {
    Pointer<Uint8>? fromBufferC;
    final toC = toPath.toNativeUtf8();
    try {
      _log.fine("[_copyMetadata] Allocating buffer with size: ${from.length}");
      fromBufferC = ffi.malloc.allocate<Uint8>(from.length);
      final fromBufferView = fromBufferC.asTypedList(from.length);
      fromBufferView.setAll(0, from);
      final result =
          _bindings.exiv2CopyMetadataFromBuffer(
            fromBufferC,
            from.length,
            toC.cast(),
            shouldCopyOrientation ? 1 : 0,
            shouldCopyThumbnail ? 1 : 0,
          ) !=
          0;
      return result;
    } finally {
      if (fromBufferC != null) {
        ffi.malloc.free(fromBufferC);
      }
      ffi.malloc.free(toC);
    }
  });
}

extension PointerListExtension<T extends NativeType> on List<Pointer<T>> {
  Pointer<Pointer<T>> toNativeArray() {
    final result = ffi.malloc.allocate<Pointer<T>>(length);
    for (final (i, e) in indexed) {
      result[i] = e;
    }
    return result;
  }
}

final _log = Logger("np_exiv2");
const String _libName = 'np_exiv2';

/// The dynamic library in which the symbols for [NpExiv2Bindings] can be found.
final DynamicLibrary _dylib = () {
  if (Platform.isMacOS || Platform.isIOS) {
    return DynamicLibrary.open('$_libName.framework/$_libName');
  }
  if (Platform.isAndroid || Platform.isLinux) {
    return DynamicLibrary.open('lib$_libName.so');
  }
  if (Platform.isWindows) {
    return DynamicLibrary.open('$_libName.dll');
  }
  throw UnsupportedError('Unknown platform: ${Platform.operatingSystem}');
}();

/// The bindings to the native functions in [_dylib].
final NpExiv2Bindings _bindings = NpExiv2Bindings(_dylib);

File: np_exiv2/src/np_exiv2.cpp
#include <cstdio>
#include <cstring>
#include <exception>
#include <exiv2/exiv2.hpp>
#include <map>
#include <string>

#include "log.h"
#include "np_exiv2.h"
#include "reader.h"

using namespace std;

#define TAG "Exiv2ReadResult"

namespace {

void convertCppType(Exiv2Metadatum *that,
                    const np_exiv2::reader::Metadatum &obj);

void convertCppType(Exiv2ReadResult *that, const np_exiv2::reader::Result &obj);

void exiv2MetadatumFree(const Exiv2Metadatum *that);

} // namespace

const Exiv2ReadResult *exiv2ReadFile(const char *path, const int is_read_xmp) {
  np_exiv2::reader::Reader reader(is_read_xmp);
  try {
    auto result = reader.readFile(path);
    if (result) {
      LOGI(TAG, "Converting result");
      auto cresult = (Exiv2ReadResult *)malloc(sizeof(Exiv2ReadResult));
      memset(cresult, 0, sizeof(Exiv2ReadResult));
      convertCppType(cresult, *result);
      LOGI(TAG, "Done");
      return cresult;
    }
  } catch (const exception &e) {
    LOGE(TAG, "Exception reading file: %s", e.what());
  } catch (...) {
    LOGE(TAG, "Exception reading file");
  }
  return nullptr;
}

const Exiv2ReadResult *exiv2ReadBuffer(const uint8_t *buffer, const size_t size,
                                       const int is_read_xmp) {
  np_exiv2::reader::Reader reader(is_read_xmp);
  try {
    auto result = reader.readBuffer(buffer, size);
    if (result) {
      LOGI(TAG, "Converting result");
      auto cresult = (Exiv2ReadResult *)malloc(sizeof(Exiv2ReadResult));
      memset(cresult, 0, sizeof(Exiv2ReadResult));
      convertCppType(cresult, *result);
      LOGI(TAG, "Done");
      return cresult;
    }
  } catch (const exception &e) {
    LOGE(TAG, "Exception reading file: %s", e.what());
  } catch (...) {
    LOGE(TAG, "Exception reading file");
  }
  return nullptr;
}

const Exiv2ReadResult *exiv2ReadHttp(const char *url, const char **header_keys,
                                     const char **header_values,
                                     const unsigned header_size,
                                     const int is_read_xmp) {
  np_exiv2::reader::Reader reader(is_read_xmp);
  map<string, string> headers;
  for (unsigned i = 0; i < header_size; ++i) {
    headers[header_keys[i]] = header_values[i];
  }
  try {
    auto result = reader.readHttp(url, headers);
    if (result) {
      LOGI(TAG, "Converting result");
      auto cresult = (Exiv2ReadResult *)malloc(sizeof(Exiv2ReadResult));
      memset(cresult, 0, sizeof(Exiv2ReadResult));
      convertCppType(cresult, *result);
      LOGI(TAG, "Done");
      return cresult;
    }
  } catch (const exception &e) {
    LOGE(TAG, "Exception reading file: %s", e.what());
  } catch (...) {
    LOGE(TAG, "Exception reading file");
  }
  return nullptr;
}

int exiv2CopyMetadataFromBuffer(const uint8_t *from_buffer,
                                const size_t from_size, const char *to,
                                const int should_copy_orientation,
                                const int should_copy_thumbnail) {
  try {
    auto src = Exiv2::ImageFactory::open(from_buffer, from_size);
    auto dst = Exiv2::ImageFactory::open(to);
    src->readMetadata();
    dst->setMetadata(*src);
    if (!should_copy_orientation) {
      try {
        auto &exif_data = dst->exifData();
        auto it = exif_data.findKey(Exiv2::ExifKey("Exif.Image.Orientation"));
        if (it != exif_data.end() && it->value().toUint32() != 1) {
          // set orientation to 1 = Horizontal (normal)
          const auto val = Exiv2::ValueType<uint16_t>(1);
          it->setValue(&val);
        }
      } catch (const exception &e) {
        LOGE(TAG, "Exception setting exif orientation: %s", e.what());
      } catch (...) {
        LOGE(TAG, "Exception setting exif orientation");
      }
    }
    if (!should_copy_thumbnail) {
      try {
        Exiv2::ExifThumb(dst->exifData()).erase();
      } catch (const exception &e) {
        LOGE(TAG, "Exception erasing exif thumbnail: %s", e.what());
      } catch (...) {
        LOGE(TAG, "Exception erasing exif thumbnail");
      }
    }
    dst->writeMetadata();
    return true;
  } catch (const exception &e) {
    LOGE(TAG, "Exception copying metadata: %s", e.what());
    return false;
  } catch (...) {
    LOGE(TAG, "Exception copying metadata");
    return false;
  }
}

void exiv2ResultFree(const Exiv2ReadResult *that) {
  if (that->iptc_data) {
    for (auto it = that->iptc_data; it != that->iptc_data + that->iptc_count;
         ++it) {
      exiv2MetadatumFree(it);
    }
    free((void *)that->iptc_data);
  }
  if (that->exif_data) {
    for (auto it = that->exif_data; it != that->exif_data + that->exif_count;
         ++it) {
      exiv2MetadatumFree(it);
    }
    free((void *)that->exif_data);
  }
  if (that->xmp_data) {
    for (auto it = that->xmp_data; it != that->xmp_data + that->xmp_count;
         ++it) {
      exiv2MetadatumFree(it);
    }
    free((void *)that->xmp_data);
  }
}

int exiv2WriteFileDateTimeOriginal(const char *path, const char *dateTime,
                                   const char *offsetTime) {
  try {
    auto image = Exiv2::ImageFactory::open(path, false);
    if (!image || !image->good()) {
      LOGE(TAG, "Failed to open image file: %s", path);
      return false;
    }
    image->readMetadata();
    auto &exifData = image->exifData();
    const Exiv2::ExifKey dateTimeKey("Exif.Photo.DateTimeOriginal");
    const Exiv2::ExifKey offsetKey("Exif.Photo.OffsetTimeOriginal");
    if (dateTime) {
      exifData[dateTimeKey.key()] = dateTime;
      if (offsetTime) {
        exifData[offsetKey.key()] = offsetTime;
      } else {
        auto it = exifData.findKey(offsetKey);
        if (it != exifData.end()) {
          exifData.erase(it);
        }
      }
    } else {
      auto it = exifData.findKey(dateTimeKey);
      if (it != exifData.end()) {
        exifData.erase(it);
      }
      it = exifData.findKey(offsetKey);
      if (it != exifData.end()) {
        exifData.erase(it);
      }
    }
    image->writeMetadata();
    return true;
  } catch (const exception &e) {
    LOGE(TAG, "Exception writing DateTimeOriginal: %s", e.what());
    return false;
  } catch (...) {
    LOGE(TAG, "Exception writing DateTimeOriginal");
    return false;
  }
}

int exiv2WriteFileGps(const char *path, const char *latitudeRef,
                      const uint32_t *latitude, const char *longitudeRef,
                      const uint32_t *longitude) {
  try {
    auto image = Exiv2::ImageFactory::open(path, false);
    if (!image || !image->good()) {
      LOGE(TAG, "Failed to open image file: %s", path);
      return false;
    }
    image->readMetadata();
    auto &exifData = image->exifData();
    const Exiv2::ExifKey gpsLatRefKey("Exif.GPSInfo.GPSLatitudeRef");
    const Exiv2::ExifKey gpsLatKey("Exif.GPSInfo.GPSLatitude");
    const Exiv2::ExifKey gpsLngRefKey("Exif.GPSInfo.GPSLongitudeRef");
    const Exiv2::ExifKey gpsLngKey("Exif.GPSInfo.GPSLongitude");

    if (latitudeRef && latitude && longitudeRef && longitude) {
      exifData[gpsLatRefKey.key()] = latitudeRef;
      Exiv2::URationalValue latVal;
      for (int i = 0; i < 3; ++i) {
        latVal.value_.push_back({latitude[i * 2], latitude[i * 2 + 1]});
      }
      exifData[gpsLatKey.key()].setValue(&latVal);

      exifData[gpsLngRefKey.key()] = longitudeRef;
      Exiv2::URationalValue lngVal;
      for (int i = 0; i < 3; ++i) {
        lngVal.value_.push_back({longitude[i * 2], longitude[i * 2 + 1]});
      }
      exifData[gpsLngKey.key()].setValue(&lngVal);
    } else {
      for (const auto &key :
           {gpsLatRefKey, gpsLatKey, gpsLngRefKey, gpsLngKey}) {
        auto it = exifData.findKey(key);
        if (it != exifData.end()) {
          exifData.erase(it);
        }
      }
    }
    image->writeMetadata();
    return true;
  } catch (const exception &e) {
    LOGE(TAG, "Exception writing GPS: %s", e.what());
    return false;
  } catch (...) {
    LOGE(TAG, "Exception writing GPS");
    return false;
  }
}

namespace {

void convertCppType(Exiv2Metadatum *that,
                    const np_exiv2::reader::Metadatum &obj) {
  auto tag_key = (char *)malloc(obj.tag_key.length() + 1);
  strcpy(tag_key, obj.tag_key.c_str());
  that->tag_key = tag_key;
  that->type_id = (Exiv2TypeId)obj.type_id;
  auto data = (uint8_t *)malloc(obj.data.size());
  memcpy(data, obj.data.data(), obj.data.size());
  that->data = data;
  that->size = obj.data.size();
  that->count = obj.count;
}

void convertCppType(Exiv2ReadResult *that,
                    const np_exiv2::reader::Result &obj) {
  that->width = obj.width;
  that->height = obj.height;

  LOGI(TAG, "Converting IPTC data");
  auto iptc_data =
      (Exiv2Metadatum *)malloc(obj.iptc_data.size() * sizeof(Exiv2Metadatum));
  auto dst_it = iptc_data;
  for (auto it = obj.iptc_data.begin(); it != obj.iptc_data.end(); ++it) {
    LOGD(TAG, "- Convert %s", it->tag_key.c_str());
    convertCppType(dst_it++, *it);
  }
  that->iptc_data = iptc_data;
  that->iptc_count = obj.iptc_data.size();

  LOGI(TAG, "Converting EXIF data");
  auto exif_data =
      (Exiv2Metadatum *)malloc(obj.exif_data.size() * sizeof(Exiv2Metadatum));
  dst_it = exif_data;
  for (auto it = obj.exif_data.begin(); it != obj.exif_data.end(); ++it) {
    LOGD(TAG, "- Convert %s", it->tag_key.c_str());
    convertCppType(dst_it++, *it);
  }
  that->exif_data = exif_data;
  that->exif_count = obj.exif_data.size();

  LOGI(TAG, "Converting XMP data");
  auto xmp_data =
      (Exiv2Metadatum *)malloc(obj.xmp_data.size() * sizeof(Exiv2Metadatum));
  dst_it = xmp_data;
  for (auto it = obj.xmp_data.begin(); it != obj.xmp_data.end(); ++it) {
    LOGD(TAG, "- Convert %s", it->tag_key.c_str());
    convertCppType(dst_it++, *it);
  }
  that->xmp_data = xmp_data;
  that->xmp_count = obj.xmp_data.size();
}

void exiv2MetadatumFree(const Exiv2Metadatum *that) {
  free((void *)that->tag_key);
  free((void *)that->data);
}

} // namespace
